File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsRootZxtunes.java
/**
 * @file
 * @brief Implementation of VfsRoot over http://zxtunes.com catalogue
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs;

import android.content.Context;
import android.net.Uri;
import android.util.SparseIntArray;

import androidx.annotation.Nullable;

import java.io.IOException;
import java.util.List;
import java.util.Locale;

import app.zxtune.R;
import app.zxtune.TimeStamp;
import app.zxtune.fs.http.MultisourceHttpProvider;
import app.zxtune.fs.zxtunes.Author;
import app.zxtune.fs.zxtunes.CachingCatalog;
import app.zxtune.fs.zxtunes.Catalog;
import app.zxtune.fs.zxtunes.Identifier;
import app.zxtune.fs.zxtunes.RemoteCatalog;
import app.zxtune.fs.zxtunes.Track;

final class VfsRootZxtunes extends StubObject implements VfsRoot {

  private static final String TAG = VfsRootZxtunes.class.getName();

  private final VfsObject parent;
  private final Context context;
  private final CachingCatalog catalog;
  private final GroupingDir[] groups;

  VfsRootZxtunes(VfsObject parent, Context context, MultisourceHttpProvider http) {
    this.parent = parent;
    this.context = context;
    this.catalog = Catalog.create(context, http);
    this.groups = new GroupingDir[]{new AuthorsDir()};
  }

  @Override
  public Uri getUri() {
    return Identifier.forRoot().build();
  }

  @Override
  public String getName() {
    return context.getString(R.string.vfs_zxtunes_root_name);
  }

  @Override
  public String getDescription() {
    return context.getString(R.string.vfs_zxtunes_root_description);
  }

  @Override
  public VfsObject getParent() {
    return parent;
  }

  @Override
  public Object getExtension(String id) {
    if (VfsExtensions.SEARCH_ENGINE.equals(id) && catalog.searchSupported()) {
      //assume root will search by authors
      return new AuthorsSearchEngine();
    } else if (VfsExtensions.ICON.equals(id)) {
      return R.drawable.ic_browser_vfs_zxtunes;
    } else {
      return super.getExtension(id);
    }
  }

  @Override
  public void enumerate(Visitor visitor) {
    for (GroupingDir group : groups) {
      visitor.onDir(group);
    }
  }

  @Override
  @Nullable
  public VfsObject resolve(Uri uri) {
    if (Identifier.isFromRoot(uri)) {
      final List<String> path = uri.getPathSegments();
      return resolve(uri, path);
    } else {
      return null;
    }
  }

  @Nullable
  private VfsObject resolve(Uri uri, List<String> path) {
    final String category = Identifier.findCategory(path);
    if (category == null) {
      return this;
    } else if (Identifier.CATEGORY_IMAGES.equals(category)) {
      return resolveImages(uri, path);
    } else {
      for (GroupingDir group : groups) {
        if (category.equals(group.getPath())) {
          return group.resolve(uri, path);
        }
      }
    }
    return null;
  }

  @Nullable
  private VfsFile resolveImages(Uri uri, List<String> path) {
    final Author author = Identifier.findAuthor(uri, path);
    if (author != null) {
      return new ImageFile(author);
    }
    return null;
  }

  private abstract class GroupingDir extends StubObject implements VfsDir {

    @Override
    public Uri getUri() {
      return Identifier.forCategory(getPath()).build();
    }

    abstract String getPath();

    abstract VfsObject resolve(Uri uri, List<String> path);
  }

  private class AuthorsDir extends GroupingDir {

    @Override
    public String getName() {
      return context.getString(R.string.vfs_zxtunes_authors_name);
    }

    @Override
    public String getDescription() {
      return context.getString(R.string.vfs_zxtunes_authors_description);
    }

    @Override
    public VfsObject getParent() {
      return VfsRootZxtunes.this;
    }

    @Override
    public Object getExtension(String id) {
      if (VfsExtensions.SEARCH_ENGINE.equals(id) && catalog.searchSupported()) {
        return new AuthorsSearchEngine();
      } else {
        return super.getExtension(id);
      }
    }

    @Override
    public void enumerate(final Visitor visitor) throws IOException {
      catalog.queryAuthors((obj) -> {
        visitor.onDir(new AuthorDir(obj));
      });
    }

    @Override
    public String getPath() {
      return Identifier.CATEGORY_AUTHORS;
    }

    @Override
    final VfsObject resolve(Uri uri, List<String> path) {
      // use plain resolving with most frequent cases first
      final Track track = Identifier.findTrack(uri, path);
      if (track != null) {
        return new TrackFile(uri, track);
      }
      final VfsObject dir = resolveDir(uri, path);
      return dir != null ? dir : this;
    }
  }

  @Nullable
  private VfsObject resolveDir(Uri uri, List<String> path) {
    final Author author = Identifier.findAuthor(uri, path);
    if (author == null) {
      return null;
    }
    final Integer date = Identifier.findDate(uri, path);
    return date != null ? new AuthorDateDir(author, date) : new AuthorDir(author);
  }

  private class AuthorDir extends StubObject implements VfsDir {

    private final Author author;

    AuthorDir(Author author) {
      this.author = author;
    }

    @Override
    public Uri getUri() {
      return Identifier.forAuthor(author).build();
    }

    @Override
    public String getName() {
      return author.getNickname();
    }

    @Override
    public String getDescription() {
      return author.getName();
    }

    @Override
    public VfsObject getParent() {
      // TODO
      return groups[0];
    }

    @Override
    public void enumerate(final Visitor visitor) throws IOException {
      final SparseIntArray dates = new SparseIntArray();
      catalog.queryAuthorTracks(author, (obj) -> {
        if (isEmptyDate(obj.getDate())) {
          final Uri uri = Identifier.forTrack(Identifier.forAuthor(author), obj).build();
          visitor.onFile(new TrackFile(uri, obj));
        } else {
          dates.put(obj.getDate(), 1 + dates.get(obj.getDate()));
        }
      });
      for (int i = 0, lim = dates.size(); i != lim; ++i) {
        visitor.onDir(new AuthorDateDir(author, dates.keyAt(i), dates.valueAt(i)));
      }
    }

    @Override
    public Object getExtension(String id) {
      if (VfsExtensions.COVER_ART_URI.equals(id)) {
        final Author obj = catalog.queryAuthor(author.getId());
        return obj != null && Boolean.TRUE == obj.getHasPhoto() ? Identifier.forPhotoOf(author) : null;
      } else {
        return super.getExtension(id);
      }
    }
  }

  private static boolean isEmptyDate(@Nullable Integer date) {
    return date == null || 0 == date;
  }

  private class AuthorDateDir extends StubObject implements VfsDir {

    private final Author author;
    private final int date;
    private final int count;

    AuthorDateDir(Author author, int date, int count) {
      this.author = author;
      this.date = date;
      this.count = count;
    }

    AuthorDateDir(Author author, int date) {
      this(author, date, 0);
    }

    @Override
    public Uri getUri() {
      return Identifier.forAuthor(author, date).build();
    }

    @Override
    public String getName() {
      return Integer.toString(date);
    }

    @Override
    public String getDescription() {
      return context.getResources().getQuantityString(R.plurals.tracks, count, count);
    }

    @Override
    public VfsObject getParent() {
      return new AuthorDir(author);
    }

    @Override
    public void enumerate(final Visitor visitor) throws IOException {
      catalog.queryAuthorTracks(author, (obj) -> {
        if (!isEmptyDate(obj.getDate()) && date == obj.getDate()) {
          final Uri uri = Identifier.forTrack(Identifier.forAuthor(author), obj).build();
          visitor.onFile(new TrackFile(uri, obj));
        }
      });
    }
  }

  private static final TimeStamp FRAME_DURATION = TimeStamp.fromMilliseconds(20);

  private class TrackFile extends StubObject implements VfsFile {

    private final Uri uri;
    private final Track module;

    TrackFile(Uri uri, Track module) {
      this.uri = uri;
      this.module = module;
    }

    @Override
    public Uri getUri() {
      return uri;
    }

    @Override
    public String getName() {
      return uri.getLastPathSegment();
    }

    @Override
    public String getDescription() {
      return module.getTitle();
    }

    @Override
    public VfsObject getParent() {
      final List<String> path = uri.getPathSegments();
      return resolveDir(uri, path);
    }

    @Override
    public Object getExtension(String id) {
      if (VfsExtensions.CACHE_PATH.equals(id)) {
        return Integer.toString(module.getId());
      } else if (VfsExtensions.DOWNLOAD_URIS.equals(id)) {
        return RemoteCatalog.getTrackUris(module.getId());
      } else if (VfsExtensions.SHARE_URL.equals(id)) {
        return getShareUrl();
      } else {
        return super.getExtension(id);
      }
    }

    @Override
    public String getSize() {
      return module.getDuration() != null ? FRAME_DURATION.multiplies(module.getDuration()).toString() : "";
    }

    @Nullable
    private String getShareUrl() {
      final Author author = Identifier.findAuthor(uri, uri.getPathSegments());
      return author != null ? String.format(Locale.US, "https://zxtunes.com/author.php?id=%d&play=%d", author.getId(), module.getId()) : null;
    }
  }

  private static class ImageFile extends StubObject implements VfsFile {
    private final Author author;

    private ImageFile(Author author) {
      this.author = author;
    }

    @Override
    public Uri getUri() {
      return Identifier.forPhotoOf(author);
    }

    @Override
    public String getName() {
      return author.getNickname();
    }

    @Override
    public VfsObject getParent() {
      return null;
    }

    @Override
    public String getSize() {
      return "";
    }

    @Override
    @Nullable
    public Object getExtension(String id) {
      if (VfsExtensions.CACHE_PATH.equals(id)) {
        return author.getId() + ".jpg";
      } else if (VfsExtensions.DOWNLOAD_URIS.equals(id)) {
        return RemoteCatalog.getImageUris(author.getId());
      } else {
        return super.getExtension(id);
      }
    }
  }

  private class AuthorsSearchEngine implements VfsExtensions.SearchEngine {

    @Override
    public void find(String query, final Visitor visitor) {
      catalog.findTracks(query, (author, track) -> {
        final Uri uri = Identifier.forTrack(Identifier.forAuthor(author), track).build();
        visitor.onFile(new TrackFile(uri, track));
      });
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsRootZxtunes.kt
/**
 * @file
 * @brief Implementation of VfsRoot over http://zxtunes.com catalogue
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.fs

import android.content.Context
import android.net.Uri
import android.util.SparseIntArray
import androidx.core.util.forEach
import app.zxtune.R
import app.zxtune.TimeStamp
import app.zxtune.fs.http.MultisourceHttpProvider
import app.zxtune.fs.zxtunes.Author
import app.zxtune.fs.zxtunes.Catalog
import app.zxtune.fs.zxtunes.Identifier
import app.zxtune.fs.zxtunes.RemoteCatalog
import app.zxtune.fs.zxtunes.Track

class VfsRootZxtunes(
    override val parent: VfsObject, private val context: Context, http: MultisourceHttpProvider
) : StubObject(), VfsRoot {
    private val catalog = Catalog.create(context, http)
    private val groups = arrayOf<GroupingDir>(AuthorsDir())

    override val uri: Uri
        get() = Identifier.forRoot().build()
    override val name
        get() = context.getString(R.string.vfs_zxtunes_root_name)
    override val description
        get() = context.getString(R.string.vfs_zxtunes_root_description)

    override fun getExtension(id: String) = when (id) {
        //assume root will search by authors
        VfsExtensions.SEARCH_ENGINE -> if (catalog.searchSupported()) AuthorsSearchEngine() else null

        VfsExtensions.ICON -> R.drawable.ic_browser_vfs_zxtunes
        else -> super.getExtension(id)
    }

    override fun enumerate(visitor: VfsDir.Visitor) = groups.forEach(visitor::onDir)

    override fun resolve(uri: Uri) = if (Identifier.isFromRoot(uri)) {
        resolve(uri, uri.pathSegments)
    } else {
        null
    }

    private fun resolve(uri: Uri, path: List<String>) =
        when (val category = Identifier.findCategory(path)) {
            null -> this
            Identifier.CATEGORY_IMAGES -> resolveImages(uri, path)
            else -> groups.find { it.path == category }?.resolve(uri, path)
        }

    private fun resolveImages(uri: Uri, path: List<String>) =
        Identifier.findAuthor(uri, path)?.let {
            ImageFile(it)
        }

    private abstract inner class GroupingDir(val path: String) : StubObject(), VfsDir {
        override val uri: Uri
            get() = Identifier.forCategory(path).build()

        abstract fun resolve(uri: Uri, path: List<String>): VfsObject?
    }

    private inner class AuthorsDir : GroupingDir(Identifier.CATEGORY_AUTHORS) {
        override val name
            get() = context.getString(R.string.vfs_zxtunes_authors_name)
        override val description
            get() = context.getString(R.string.vfs_zxtunes_authors_description)
        override val parent
            get() = this@VfsRootZxtunes

        override fun getExtension(id: String) = when (id) {
            VfsExtensions.SEARCH_ENGINE -> if (catalog.searchSupported()) AuthorsSearchEngine()
            else null

            else -> super.getExtension(id)
        }

        override fun enumerate(visitor: VfsDir.Visitor) = catalog.queryAuthors { obj ->
            visitor.onDir(AuthorDir(obj))
        }

        // use plain resolving with most frequent cases first
        override fun resolve(uri: Uri, path: List<String>) = Identifier.findTrack(uri, path)?.let {
            TrackFile(uri, it)
        } ?: resolveDir(uri, path) ?: this

    }

    private fun resolveDir(uri: Uri, path: List<String>) =
        Identifier.findAuthor(uri, path)?.let { author ->
            Identifier.findDate(uri, path)?.let { date ->
                AuthorDateDir(author, date, 0)
            } ?: AuthorDir(author)
        }

    private inner class AuthorDir(private val author: Author) : StubObject(), VfsDir {
        override val uri: Uri
            get() = Identifier.forAuthor(author).build()
        override val name
            get() = author.nickname
        override val description
            get() = author.name
        override val parent
            get() = groups[0] // TODO

        override fun enumerate(visitor: VfsDir.Visitor) = SparseIntArray().apply {
            catalog.queryAuthorTracks(author) { obj ->
                if (obj.date == null || obj.date == 0) {
                    val uri = Identifier.forTrack(Identifier.forAuthor(author), obj).build()
                    visitor.onFile(TrackFile(uri, obj))
                } else {
                    put(obj.date, 1 + get(obj.date))
                }
            }
        }.forEach { key, value ->
            visitor.onDir(AuthorDateDir(author, key, value))
        }

        override fun getExtension(id: String) = when (id) {
            VfsExtensions.COVER_ART_URI -> catalog.queryAuthor(author.id)
                ?.takeIf { true == it.hasPhoto }?.let { Identifier.forPhotoOf(it) }
            // assume called on enumerated object with full Author
            VfsExtensions.ICON_URI -> author.takeIf { true == it.hasPhoto }?.let {
                Identifier.forPhotoOf(it)
            }
            else -> super.getExtension(id)
        }
    }

    private inner class AuthorDateDir(
        private val author: Author, private val date: Int, private val count: Int
    ) : StubObject(), VfsDir {
        override val uri: Uri
            get() = Identifier.forAuthor(author, date).build()
        override val name
            get() = date.toString()
        override val description
            get() = context.resources.getQuantityString(R.plurals.tracks, count, count)
        override val parent
            get() = AuthorDir(author)

        override fun enumerate(visitor: VfsDir.Visitor) = catalog.queryAuthorTracks(author) { obj ->
            if (date == obj.date) {
                val uri = Identifier.forTrack(Identifier.forAuthor(author), obj).build()
                visitor.onFile(TrackFile(uri, obj))
            }
        }
    }

    private inner class TrackFile(override val uri: Uri, private val module: Track) : StubObject(),
        VfsFile {
        override val name
            get() = requireNotNull(uri.lastPathSegment)
        override val description
            get() = module.title
        override val parent
            get() = resolveDir(uri, uri.pathSegments)

        override fun getExtension(id: String) = when (id) {
            VfsExtensions.CACHE_PATH -> module.id.toString()
            VfsExtensions.DOWNLOAD_URIS -> RemoteCatalog.getTrackUris(module.id)
            VfsExtensions.SHARE_URL -> shareUrl
            else -> super.getExtension(id)
        }

        override val size
            get() = module.duration?.let {
                FRAME_DURATION.multiplies(it.toLong()).toString()
            } ?: ""

        val shareUrl
            get() = Identifier.findAuthor(uri, uri.pathSegments)?.let { author ->
                "http://zxtunes.com/author.php?id=${author.id}&play=${module.id}"
            }
    }

    private class ImageFile(
        private val author: Author,
        override val parent: VfsObject? = null,
        override val size: String = ""
    ) : StubObject(), VfsFile {
        override val uri: Uri
            get() = Identifier.forPhotoOf(author)
        override val name
            get() = author.nickname

        override fun getExtension(id: String) = when (id) {
            VfsExtensions.DOWNLOAD_URIS -> RemoteCatalog.getImageUris(author.id)
            else -> super.getExtension(id)
        }
    }

    private inner class AuthorsSearchEngine : VfsExtensions.SearchEngine {
        override fun find(query: String, visitor: VfsExtensions.SearchEngine.Visitor) =
            catalog.findTracks(query) { author, track ->
                val uri = Identifier.forTrack(Identifier.forAuthor(author), track).build()
                visitor.onFile(TrackFile(uri, track))
            }
    }

    companion object {
        private val FRAME_DURATION = TimeStamp.fromMilliseconds(20)
    }
}
